feat(traverse): record current block scope

overlookmotel opened the pull request:
Record "block" scope ID along with "hoist" scope ID in `Traverse`.

"Block" scope is the scope where a `let` statement would be inserted above current position in AST.

Block scope and current scope differ from each other inside classes. For example, if want to create a `let` temp var for `foo()` or `bar()` in example below, should use block scope not current scope. Current scope is the class itself, but the `let` statement will be inserted *outside* the class.

```js
class C {
  [foo()]: bar();
}
```

All transforms which create `let` bindings should use block scope not current scope. We should add `VarDeclarationsStore::insert_let` method which uses block scope, to accompany `insert_var` (which uses hoist scope).

This is implemented in a rather hacky way, and we should improve it later. Notably, we're not considering `for` statements as block scopes because we currently have no way to insert `let` statements into them if they don't have a body block.

overlookmotel commented:
> [!WARNING]
> <b>This pull request is not mergeable via GitHub because a downstack PR is open. Once all requirements are satisfied, merge this PR as a stack <a href="https://app.graphite.dev/github/pr/oxc-project/oxc/8007?utm_source=stack-comment-downstack-mergeability-warning" >on Graphite</a>.</b>
> <a href="https://graphite.dev/docs/merge-pull-requests">Learn more</a>

* **#7982** <a href="https://app.graphite.dev/github/pr/oxc-project/oxc/7982?utm_source=stack-comment-icon" target="_blank"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="10px" height="10px"/></a>
* **#8007** <a href="https://app.graphite.dev/github/pr/oxc-project/oxc/8007?utm_source=stack-comment-icon" target="_blank"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="10px" height="10px"/></a> 👈 <a href="https://app.graphite.dev/github/pr/oxc-project/oxc/8007?utm_source=stack-comment-view-in-graphite" target="_blank">(View in Graphite)</a>
* **#8001** <a href="https://app.graphite.dev/github/pr/oxc-project/oxc/8001?utm_source=stack-comment-icon" target="_blank"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="10px" height="10px"/></a>
* **#7995** <a href="https://app.graphite.dev/github/pr/oxc-project/oxc/7995?utm_source=stack-comment-icon" target="_blank"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="10px" height="10px"/></a>
* **#7998** <a href="https://app.graphite.dev/github/pr/oxc-project/oxc/7998?utm_source=stack-comment-icon" target="_blank"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="10px" height="10px"/></a>
* `main`
<hr />

## How to use the Graphite Merge Queue

Add either label to this PR to merge it via the merge queue:
- <kbd>0-merge</kbd> - adds this PR to the back of the merge queue
- <kbd>hotfix</kbd> - for urgent hot fixes, skip the queue and merge this PR next

You must have a Graphite account in order to use the merge queue. Sign up using [this link](https://app.graphite.dev/invite/github/oxc-project?ref=merge-queue-instructions-in-stack-comment&prId=8007).

<h2></h2>
<sub>

An organization admin has enabled the [Graphite Merge Queue](https://graphite.dev/docs/graphite-merge-queue) in this repository.

</sub>

<sub>

Please do not merge from GitHub as this will restart CI on PRs being processed by the merge queue.

</sub>

<h2></h2>

This stack of pull requests is managed by <a href="https://graphite.dev?utm-source=stack-comment"><b>Graphite</b></a>. Learn more about <a href="https://stacking.dev/?utm_source=stack-comment">stacking</a>.
<!-- Current dependencies on/for this PR: -->

Dunqing commented:
I have taken a look at the whole stack. From the usages, it seems we can get `current_block_scope` by looking up the parent scope of `current_scope_id` because when you want to use `current_block_scope`, you are always at the `class` scope, so you only need to look up once and store it somewhere. That I understand correctly?

To clarify, I am not rejecting this API at all; on the contrary, I am happy to add various APIs in `Traverse` and `TransformerCtx` to solve annoying duplicating code. So If you find that can be fixed or refactored in other places by it, tell me!

overlookmotel commented:
> it seems we can get `current_block_scope` by looking up the parent scope of `current_scope_id` because when you want to use `current_block_scope`, you are always at the `class` scope, so you only need to look up once and store it somewhere. That I understand correctly?

That's what I thought but actually no!

This gem was in one of Babel's tests:

```js
class Outer extends class Inner {
  [foo()] = 123
} {}
```

So when you're in the inner class, is *not* where a `let` binding gets inserted - the parent scope is the scope of the outer class!

Dunqing commented:
> > it seems we can get `current_block_scope` by looking up the parent scope of `current_scope_id` because when you want to use `current_block_scope`, you are always at the `class` scope, so you only need to look up once and store it somewhere. That I understand correctly?
> 
> That's what I thought but actually no!
> 
> This gem was in one of Babel's tests:
> 
> ```js
> class Outer extends class Inner {
>   [foo()] = 123;
> } {}
> ```
> 
> So when you're in the inner class, the parent scope is _not_ where a `let` binding gets inserted - the parent scope is the scope of the _outer_ class!

Aha, I see the problem, fu** JavaScript!

overlookmotel committed:
feat(traverse): record current block scope (#8007)

Record "block" scope ID along with "hoist" scope ID in `Traverse`.

"Block" scope is the scope where a `let` statement would be inserted above current position in AST.

Block scope and current scope differ from each other inside classes. For example, if want to create a `let` temp var for `foo()` or `bar()` in example below, should use block scope not current scope. Current scope is the class itself, but the `let` statement will be inserted *outside* the class.

```js
class C {
  [foo()]: bar();
}
```

All transforms which create `let` bindings should use block scope not current scope. We should add `VarDeclarationsStore::insert_let` method which uses block scope, to accompany `insert_var` (which uses hoist scope).

This is implemented in a rather hacky way, and we should improve it later. Notably, we're not considering `for` statements as block scopes because we currently have no way to insert `let` statements into them if they don't have a body block.
